### PR TITLE
Planning tests, comments and minor improvements for boundary cases.

### PR DESCRIPTION
**Pull request by samagra14**
This pull request finishes the implementation of Chapter 10 complete with comments and test cases. This PR also improves the existing implementation for certain boundary cases. This pull request also adds a small `permute` package for combinatorial calculations.

### CODE REVIEW

**Comment by samagra14** [comment]
@norvig Please have a look here.

File: aima-core/src/main/java/aima/core/logic/planning/ActionSchema.java [hunk 1]
```diff
@@ -1,24 +1,43 @@
...
+ *
+ * @author samagra
+ */
 public class ActionSchema {
-    List<Term> variables;
-    List<Literal> precondition; //treated as a conjucntion of fluents
-    List<Literal> effects;//treated as a conjunction of fluents
+    List<Term> variables;// list of variables
+    List<Literal> precondition; //PRECONDITION: treated as a conjunction of fluents
+    List<Literal> effects;//EFFECT: treated as a conjunction of fluents
```
Comment (norvig) [inline aima-core/src/main/java/aima/core/logic/planning/ActionSchema.java:26]: Probably there should be a space between the ";" and the "//"

File: aima-core/src/main/java/aima/core/logic/planning/GraphPlanAlgorithm.java [hunk 2]
```diff
@@ -77,52 +130,115 @@
...
-                solution.add(setToBeTaken);
-                for (ActionSchema action :
-                        setToBeTaken) {
-                    for (Literal literal :
-                            action.getPrecondition()) {
-                        if (!goalsCurr.contains(literal)) {
-                            goalsCurr.add(literal);
-                        }
+            solution.add(setToBeTaken);
+            for (ActionSchema action :
```
Comment (norvig) [inline aima-core/src/main/java/aima/core/logic/planning/GraphPlanAlgorithm.java:144]: I think these can fit on one line.

File: aima-core/src/main/java/aima/core/logic/planning/Level.java [hunk 3]
```diff
@@ -233,8 +268,8 @@ private void calculateNextLinks() {
         } else if (levelObjects.get(0) instanceof ActionSchema) {
             for (Object action :
                     levelObjects) {
-                Literal[] effects = (Literal[]) ((ActionSchema) action).getEffects().toArray();
-                nextLinks.put(action, Arrays.asList((Object[]) effects));
+                Object[] effects =  ((ActionSchema) action).getEffects().toArray();
```
Comment (norvig) [inline aima-core/src/main/java/aima/core/logic/planning/Level.java:271]: Why isn't this a `Literal[]` ?

File: aima-core/src/main/java/aima/core/logic/planning/Utils.java [hunk 4]
```diff
@@ -9,10 +9,16 @@
...
+ * Utils for parsing planning problems.
+ *
+ * @author samagra
+ */
 public class Utils {
+
     public static List<Literal> parse(String s) {
         if (s.isEmpty())
-            return null;
+            return new ArrayList<>();
```
Comment (norvig) [inline aima-core/src/main/java/aima/core/logic/planning/Utils.java:21]: Always better to return an empty ArrayList than null.

File: aima-core/src/main/java/aima/core/logic/planning/Level.java (hunk 3 above)
Comment (samagra14) [reply aima-core/src/main/java/aima/core/logic/planning/Level.java:271]: @norvig 
Because of the hashmap on line 272. `nextLinks` require a List of objects to be added as values. This is done so that levels consisting of both actionSchemas as well as literals can be handled in a similar way.

File: aima-core/src/main/java/aima/core/logic/planning/Utils.java (hunk 4 above)
Comment (samagra14) [reply aima-core/src/main/java/aima/core/logic/planning/Utils.java:21]: :sweat_smile:

File: aima-core/src/main/java/aima/core/logic/planning/Level.java (hunk 3 above)
Comment (norvig) [reply aima-core/src/main/java/aima/core/logic/planning/Level.java:271]: Got it, thanks.
